### ISSUE TITLE
Failing unification slower in 8.12.1 than in 8.11.2

### ISSUE THREAD

**Issue by Janno** [coq/coq#13419]
#### Description of the problem

The call to `refine` in the code below takes almost no time (~5 ms incl. printing the error) to fail with Coq 8.11.2. With 8.12.1 it takes forever.

```coq
From Coq.ZArith Require Import BinIntDef.
From Coq Require Import List String.
Import BinNums.
Import ListNotations.
Import Z.
Open Scope Z_scope.

Goal forall a : Z,
  @Logic.eq Z
    ((fix get (ls : list Z) : Z :=
        match ls return Z with
        | nil => Zpos (xO (xO (xO (xO (xO (xO xH))))))
        | cons l ls0 => match testbit a l return Z with
                        | true => l
                        | false => get ls0
                        end
        end) (@map nat Z of_nat (seq O (BinPos.Pos.to_nat (xO (xO (xO (xO (xO (xO xH))))))))))
    match eqb a Z0 return Z with
    | true => sub (pow (Zpos (xO xH)) (Zpos (xO (xO (xO (xO (xO (xO xH)))))))) (Zpos xH)
    | false =>
        (fix get (ls : list Z) : Z :=
           match ls return Z with
           | nil => Zpos (xO (xO (xO (xO (xO (xO xH))))))
           | cons l ls0 => match testbit a l return Z with
                           | true => l
                           | false => get ls0
                           end
           end) (@map nat Z of_nat (seq O (BinNat.N.to_nat (Npos (xO (xO (xO (xO (xO (xO xH))))))))))
    end

.
  intros.
  Time Fail refine (@Logic.eq_refl _ _).
```

#### Coq Version
8.12.1

**Comment by Janno** [comment]
I can and will bisect this.

**Comment by Zimmi48** [comment]
FWIW, this already takes forever in 8.12.0.

**Comment by Janno** [comment]
Got a result!

```
df8df4637dfb4106854554cc2ac94b4fdd565e80 is the first bad commit
commit df8df4637dfb4106854554cc2ac94b4fdd565e80
Author: Hugo Herbelin <Hugo.Herbelin@inria.fr>
Date:   Sun Apr 19 13:00:09 2020 +0200

    Fixing #11903: Fixpoints not truly recursive in standard library.
    
    There was also a non truly recursive in the doc.

 doc/sphinx/proof-engine/ssreflect-proof-language.rst | 2 +-
 theories/FSets/FMapAVL.v                             | 4 ++--
 theories/Init/Decimal.v                              | 2 +-
 theories/NArith/BinNatDef.v                          | 4 ++--
 theories/ZArith/BinIntDef.v                          | 2 +-
 theories/extraction/ExtrOcamlBigIntConv.v            | 4 ++--
 theories/extraction/ExtrOcamlIntConv.v               | 4 ++--
 7 files changed, 11 insertions(+), 11 deletions(-)
```

**Comment by gares** [comment]
Commit df8df4637dfb4106854554cc2ac94b4fdd565e80 turns `Fix` nodes to `Const`. IIRC evarconv "synchronizes" in different ways fix and const. CC @herbelin

**Comment by gares** [comment]
@Janno it would help to know how you ended up with that example. While it is a perfect test case for this regression, it looks like there used to be a readable goal and then someone did random unfoldings and called reflexivity. How did you end up in that situation exactly?

**Comment by SkySkimmer** [comment]
I think this is kernel conversion (ccnv), not evarconv (the 2 sides of the equality get converted and have no holes so delegated to the kernel)

**Comment by Zimmi48** [comment]
That would explain why on Zulip @Janno said all tactics (including `apply` and `apply:`) were affected.

**Comment by Janno** [comment]
> @Janno it would help to know how you ended up with that example. While it is a perfect test case for this regression, it looks like there used to be a readable goal and then someone did random unfoldings and called reflexivity. How did you end up in that situation exactly?

This problem arises in our automation for separation logic proofs. Part of that is cancelling terms in the premises and the conclusion and this particular unification attempt is the result of trying to cancel two terms that are *almost* equal. The actual terms are more complex but the goal in this issue corresponds to the subtree of unification that hangs. If you want something  nicer you can imagine the unification problem as
```coq
  builtins.trailing_zeros W64 value =?= if value =? 0 then 2 ^ 64 - 1 else builtins.trailing_zeros W64 value
```
where `trailing_zeros` is defined in
https://github.com/bedrocksystems/cpp2v/blob/master/theories/lang/cpp/semantics/builtins.v#L35
and `=?` is `Z.eqb`.

> That would explain why on Zulip @Janno said all tactics (including apply and apply:) were affected.

This is indeed my attempt at minimizing the problem I mentioned in the zulip chat. Sadly I lack the debugging tools to really figure out what's going on with the order of unification. As mentioned in zulip the actual unification problem is much bigger but trivially unsatisfiable if arguments are unified left-to-right. But I cannot even figure out what order is used by Coq. (I would have to read the sources..)

**Comment by gares** [comment]
@SkySkimmer I agree that evar-free terms are converted using the kernel, and the kernel goes right-to-left. But this does not explain the change in behavior from 8.11 (the kernel was also right-to-left back then as well).

Maybe 8.12 solves an evar which was not solved in 8.11 (different postponements, candidates...) and passes the problem to the kernel, while 8.11 was not?

**Comment by Janno** [comment]
I think the terms are free of evars in 8.11 and 8.12.

**Comment by gares** [comment]
BTW, keeping around a term like `2 ^ 64 - 1` without locking it seems like climbing without a rope...

**Comment by SkySkimmer** [comment]
When `eqb` (on Z) is defined as a fixpoint (the first argument selected as the decreasing argument arbitrarily) (8.11 version), the kernel reduces the right side to 
~~~
Rel 1 | ZFix ((fix eqb ...), []) ; Zapp [0]; Zcase (on the bool returned by eqb)
~~~
When it's a regular match (8.12 version) we get instead 
~~~
Rel 1 | Zcase (eqb contents) | Zcase (on the bool returned by eqb)
~~~

The left side reduces to
~~~
Rel 1 | Zcase (contents of odd, from "testbit a 0") | Zcase (the "match testbit a 0", 2nd branch has a big recursive case (from 1 to lots))
~~~

The heads are always the same, so we compare the stacks. In 8.11 `compare_stack_shape` sees that `zfix` is only on one side and bails. In 8.12 we compare the outer matches first, this basically compares `get (map of_nat (seq 1 lots))` and `get (map of_nat (seq 0 lots))` which takes forever to fail.

In other words, when comparing blocked matches
~~~
match (match head args return p1 with br1 end) return p2 with br2 end
==
match (match head args' return p1' with br1' end) return p2' with br2' end
~~~
we do in order
- check that the inductives of the match scrutinees line up, in the example this is `bool` for the outer match and `Z` for the inner match so we don't get to bail out
- `p2 == p2'` (in the example both `Z`)
- `br2 == br2'` (slow in the example) (in reverse order of constructors, we finally fail when checking the branches for `true`)
- `p1 == p1'` (in the example, never runs)
- `br1 == br1'` (idem)
- `args == args'` (idem) (right to left)

Notice the "in reverse order of constructors", indeed if we swap the branch contents for the match on `testbit` on both sides we fail fast. Changing this order may be OK but the symmetric example will be slow so not really a solution.

I'm not sure why we compare the outer stack elements first. Here it's slower, and in general it seems incorrect: the outer matches may be on scrutinees at different parameters of the same inductive until we have converted the scrutinees, so our typing assumptions may fail.
Indeed I checked and changing this order causes immediate failure, and the stdlib passes so I'll make a PR and see how CI/bench go.

**Comment by SkySkimmer** [comment]
https://github.com/coq/coq/pull/13424

PS: if you do
~~~coq

Definition foo a :=
  eq_refl :
    ((fix get (ls : list Z) : Z :=
        match ls return Z with
        | nil => Zpos (xO (xO (xO (xO (xO (xO xH))))))
        | cons l ls0 => match testbit a l return Z with
                        | true => l
                        | false => get ls0
                        end
        end) (@map nat Z of_nat (seq O (BinPos.Pos.to_nat (xO (xO (xO (xO (xO (xO xH))))))))))
    = _.
Print foo.
~~~
the right side generated by unification (evarconv) is pretty ugly, it seems to have done some ad-hoc reduction for some reason.

**Comment by Blaisorblade** [comment]
> When eqb (on Z) is defined as a fixpoint (the first argument selected as the decreasing argument arbitrarily) (8.11 version), the kernel reduces the right side to

> When it's a regular match (8.12 version) we get instead

Ah, so fixing the 8.12 warning against useless fixpoints is API-breaking here (as well?), and one should expect knock-on effects for each such fixed warning. Maybe one should add a warning next to mentions of https://github.com/coq/coq/pull/12121 in the changelog—or is that overkill?

**Comment by gares** [comment]
Yes and no. To me the problem is that there is a large computable term hanging around (a bomb).
This makes your scripts very fragile to almost any operation. Even a tiny call to simpl or trivial
can just take forever. Running automation is like walking on a minefield.

My understanding here is that we were lucky in this case, or better unlucky in removing a useless fix. But as we can see the PR trying to change conversion to address this, its not obvious (it's faster here, but not there.. https://gitlab.com/coq/coq/-/jobs/861521154). The only solution to this problem is to write scripts that are less incline to explode (this is why I was asking were the term comes from). We may lack tools to easily block computation, but fiddling with the heuristics is not going to cut it, IMO.

**Comment by ppedrot** [comment]
So, the conversation is gearing towards an INVALID / WONTFIX?

**Comment by herbelin** [comment]
By the way, if it is not already the case, that' ll be good to have a place which collects all tricks already tested in conversion which failed for a reason or another:
- first-to-last/last-to-first arguments comparison
- testing different conversion strategies in parallel
- ordering or weighting the constants to provide a heuristic on which term, on left or on right side, and which argument to unfold first
- ...

Incidentally, may `Opaque` on `^` have an impact here?

**Comment by ppedrot** [comment]
There is this: https://github.com/coq/coq/wiki/Performance-experiments.

**Comment by herbelin** [comment]
> There is this: https://github.com/coq/coq/wiki/Performance-experiments.

Thanks!

Then, do we have an idea of the form of problems which work better with one strategy and the form of problems which work better with another?

**Comment by herbelin** [comment]
@gares: Do you know the paper "Sharing Equality is Linear" by Condoluci, Accattoli, and Sacerdoti Coen and if yes, if some of the ideas are applicable?

**Comment by herbelin** [comment]
> #13424
> 
> PS: if you do
> 
> ```coq
> Definition foo a :=
>   eq_refl :
>     ((fix get (ls : list Z) : Z :=
>         match ls return Z with
>         | nil => Zpos (xO (xO (xO (xO (xO (xO xH))))))
>         | cons l ls0 => match testbit a l return Z with
>                         | true => l
>                         | false => get ls0
>                         end
>         end) (@map nat Z of_nat (seq O (BinPos.Pos.to_nat (xO (xO (xO (xO (xO (xO xH))))))))))
>     = _.
> Print foo.
> ```
> 
> the right side generated by unification (evarconv) is pretty ugly, it seems to have done some ad-hoc reduction for some reason.

Don't know what to think. With "global fixpoints", exposing such a `get` would not have to arrive. Now, the evaluable `fix` being there, it goes through `whd_betaiota_deltazeta_for_iota_state` in `evarconv` which is probably the culprit for expanding the `map` and triggering the reduction of the `fix`. This `whd_betaiota_deltazeta_for_iota_state` is there for a very long time and I never really tried to play with it and see if something better could be done. In the case of a reflexivity, it is indeed unexpected that a reduction is used.

Somehow, the "good" notion of reduction would anyway be to `betaiota` only _after_ a substitution, so that if a term has beta-iota-fix-cofix redexes on purpose, they are preserved. If we do it in evarconv and various other places, it is so that the term looks a priori "nicer" but it is too late: it is primarily the redexes generated by a substitution which contributes to make a term less nicer than what the user provided.

**Comment by gares** [comment]
I know something. And I know they are working with Bruno on the topic. The main thing that clashes is that these works do full reduction, not whd.... Time will tell

**Comment by herbelin** [comment]
To continue with questions on this topic, the current problem is somehow simple, i.e. almost of the form `if a =? 0 then FOO else BIG` =?= `BIG` [*]. Because of one term being a subterm of the other, there is 0-chance that reducing the right-hand side gives anything useful. The only hope is that reducing on the left either reduces the non trivial cycle into an identity or eventually discards the circularity (forcing then to reduce `BIG`).

I don't know how much this can be turned into a general heuristics. Should we think at heuristics to recognize what is liable to be a "big" term?

[*] I say almost because there is small redex `BinNat.N.to_nat (Npos _)` on the way but we could imagine, in a system with more reasoning modulo equivalence, or looking at shapes only, that this small redex becomes transparent.

For the record, an example I met in the past (in LinAlg contribution) was of the form `proj1 (proj1 ((t,u),w))` = `proj1 (t,v)` with `t` a big record full of proofs. Because the head constant was the same, `Opaque` was not helping. Here is a reduced example:
```
Fixpoint ack n :=
  match n with 0 => S | S n => fix ackSn m := match m with 0 => ack n 1 | S m => ack n (ackSn m) end end.

Goal fst ((ack 4 4, 1)) = fst (fst ((ack 4 4, 2), 3)).
reflexivity. (* hangs *)
```

This incidentally suggests that projections should not be treated as constants but as destructors, i.e. that the lazy weak-head normal form of `fst (t,u)` is not `< fst | (t,u) · TOP >` but `< (t,u) | fst · TOP >` (even if not a primitive projection).

Another remark is that even after having reached the two nested `match` shown by @SkySkimmer (even if intuitively too late), this nesting computationally corresponds to a commutative cut, so it is not absurd a priori to treat it as a single `match` and to consider that #13424 is worth further explorations, in addition of course to other explorations.

As another exploration precisely, at a time where everyone is talking machine learning and probabilistic programming, why not to check several problems in parallel? Or, at least to use extra-logical heuristics, such as the size or number of constants in a branch to decide which one to compare first?

Not to tell that I have a magic solution, but certainly to tell that it is a topic where we shouldn't refrain ourselves to brainstorm and to share the examples that fail so that we can continue learning from these failures.

**Comment by herbelin** [comment]
> I know something. And I know they are working with Bruno on the topic. The main thing that clashes is that these works do full reduction, not whd.... Time will tell

Thanks, that would be great if something comes out of it.

**Comment by ejgallego** [comment]
What should we do with this issue? I'm not an expert but I kind of share Enrico's view.

IMHO `locking/unlocking` should be advertised more promiently to users.
